src/protojson/JobReadyMsg.cc

```diff
@@ -0,0 +1,116 @@
+
+// Qserv headers
+#include "http/Client.h"
+#include "http/MetaModule.h"
+#include "http/RequestBodyJSON.h"
+#include "util/common.h"
+#include "util/TimeUtils.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+using namespace std;
+using namespace nlohmann;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.protojson.JobReadyMsg");
+}  // namespace
+
+namespace lsst::qserv::protojson {
+
+JobReadyMsg::Ptr JobReadyMsg::create(std::string const& replicationInstanceId,
+                                     std::string const& replicationAuthKey, std::string const& workerIdStr,
+                                     std::string const& czarName, CzarIdType czarId, QueryId queryId,
+                                     UberJobId uberJobId, std::string const& fileUrl, uint64_t rowCount,
+                                     uint64_t fileSize) {
+    auto jrMsg = Ptr(new JobReadyMsg(replicationInstanceId, replicationAuthKey));
+    jrMsg->_workerId = workerIdStr;
+    jrMsg->_czarName = czarName;
+    jrMsg->_czarId = czarId;
+    jrMsg->_queryId = queryId;
+    jrMsg->_uberJobId = uberJobId;
+    jrMsg->_fileUrl = fileUrl;
+    jrMsg->_rowCount = rowCount;
+    jrMsg->_fileSize = fileSize;
+    return jrMsg;
+}
+
+JobReadyMsg::Ptr JobReadyMsg::createFromJson(nlohmann::json const& jsWReq,
+                                             std::string const& replicationInstanceId,
+                                             std::string const& replicationAuthKey) {
+    string const fName("JobReadyMsg::createFromJson");
+    LOGS(_log, LOG_LVL_DEBUG, fName);
+    try {
+        if (jsWReq["version"] != http::MetaModule::version) {
+            LOGS(_log, LOG_LVL_ERROR, fName << " bad version");
+            return nullptr;
+        }
+
+        // Presumably, if these were wrong, it wouldn't have gotten this far.
+        auto repliInstId = http::RequestBodyJSON::required<string>(jsWReq, "instance_id");
+        auto repliAuthKey = http::RequestBodyJSON::required<string>(jsWReq, "auth_key");
+
+        auto jrMsg = create(repliInstId, repliAuthKey);
+
+        jrMsg->_workerId = http::RequestBodyJSON::required<string>(jsWReq, "workerid");
+        jrMsg->_czarName = http::RequestBodyJSON::required<string>(jsWReq, "czar");
+        jrMsg->_czarId = http::RequestBodyJSON::required<CzarIdType>(jsWReq, "czarid");
+        jrMsg->_queryId = http::RequestBodyJSON::required<QueryId>(jsWReq, "queryid");
+        jrMsg->_uberJobId = http::RequestBodyJSON::required<UberJobId>(jsWReq, "uberjobid");
+        jrMsg->_fileUrl = http::RequestBodyJSON::required<string>(jsWReq, "fileUrl");
+        jrMsg->_rowCount = http::RequestBodyJSON::required<uint64_t>(jsWReq, "rowCount");
+        jrMsg->_fileSize = http::RequestBodyJSON::required<uint64_t>(jsWReq, "fileSize");
+        return jrMsg;
+    } catch (invalid_argument const& exc) {
+        LOGS(_log, LOG_LVL_ERROR, string("JobReadyMsg::createJson invalid ") << exc.what());
+    }
+    return nullptr;
+}
```

Comment:
This is not a good code - the factory method should be simple. What you are doing here has to be done by another (overloaded) c-tor of the class. Something like:
```
JobReadyMsg::JobReadyMsg(nlohmann::json const& jsWReq, ... );
```
So your factory method would simply do:
```
JobReadyMsg::Ptr JobReadyMsg::createFromJson(nlohmann::json const& jsWReq, ...) {
    try {
       return shared_ptr<JobReadyMsg>(new JobReadyMsg(jsWReq, ...);
    } catch (invalid_argument const& exc) {
        LOGS(_log, LOG_LVL_ERROR, string("JobReadyMsg::createJson invalid ") << exc.what());
            return nullptr;
    }
}
```
The factory method should not be aware too much about the internals of the class construction and parameter validation which is normally a responsibility of constructors.

Reply:
I don't really get your argument here. The factory method is a member of the class and keeps the internals  encapsulated. I'm changing the code, but the reason that a factory method must be incredibly simple isn't clear to me (and this really isn't complicated). If the factory function wasn't a member of the class, I would agree due to breaking encapsulation, but that isn't the case.

I usually think constructors need to be simple since you're usually trying to get everything defined in the block before the constructor body.

Reply:
The basic rule for what one can do in the factory method and what can not is - the factory method should do what can't be done in the constructors.